util/fiber_socket_base.h

```diff
@@ -124,7 +125,11 @@ class FiberSocketBase : public io::Sink,
   virtual void CancelOnErrorCb() = 0;
 
   struct RecvNotification {
-    // TODO: move ProvidedBuffer functionality here and remove RecvProvided interface.
+    // For pure epoll notifications, this contains a monostate.
+    // For iouring/bufring notifications, it contains either the positive error code,
+    // or the received data buffer. In latter case, the callback must consume the data before
+    // returning.
+    std::variant<std::monostate, int, io::MutableBytes> read_result;
```

Comment:
MutableBytes is already 16 bytes, so an error_code seems nicer than a tricky unix-c-stdlib negative int error code